### DISCUSSION TITLE
Unable to access the traffic tile after map-matching

### DISCUSSION THREAD

**Discussion by jjeffchoi** [valhalla/valhalla#4767] [Q&A]
I'm attempting to match global coordinates with traffic tile's edges, but having some difficulty accessing the tiles.

First, I'm using the map-matching feature of Valhalla to find the `GraphId`'s corresponding to the provided global coordinates:

```cpp
// prepare data to create measurements
const float default_gps_accuracy = mapmatcher->config().emission_cost.gps_accuracy_meters,
      default_search_radius = mapmatcher->config().candidate_search.search_radius_meters;

std::vector<std::tuple<double, double, float, float, double, valhalla::baldr::Location::StopType>> data;
data.reserve(coordinates.size());
for (const Coordinate& coordinate: coordinates) {
    data.emplace_back(std::make_tuple(
                coordinate.lng,
                coordinate.lat,
                default_gps_accuracy,
                default_search_radius,
                1718210310, // hardcoded for now just for testing
                valhalla::baldr::Location::StopType::VIA
                ));
}

if (data.empty()) {
    std::cout << "No coordinate data to map match" << std::endl;
    return std::vector<valhalla::baldr::GraphId>();
}

std::vector<valhalla::meili::Measurement> measurements = CreateMeasurements(data);

if (measurements.empty()) {
    std::cout << "No measurements to map match" << std::endl;
    return std::vector<valhalla::baldr::GraphId>();
}

std::vector<valhalla::meili::MatchResult> results = mapmatcher->OfflineMatch(measurements).front().results;
```

Then, I'm passing `GraphId` from `MatchResult` as a parameter to `customize_live_traffic_data` to find a specific edge in a traffic tile:
```cpp
void customize_live_traffic_data(const boost::property_tree::ptree& config,
                                 const uint32_t& constant_encoded_speed,
                                 const valhalla::baldr::GraphId& edge_id) {
  const auto memory =
      std::make_shared<MMap>(config.get<std::string>("mjolnir.traffic_extract").c_str());

  mtar_t tar;
  tar.pos = 0;
  tar.stream = memory->data;
  tar.read = [](mtar_t* tar, void* data, unsigned size) -> int {
    memcpy(data, reinterpret_cast<char*>(tar->stream) + tar->pos, size);
    return MTAR_ESUCCESS;
  };
  tar.write = [](mtar_t* tar, const void* data, unsigned size) -> int {
    memcpy(reinterpret_cast<char*>(tar->stream) + tar->pos, data, size);
    return MTAR_ESUCCESS;
  };
  tar.seek = [](mtar_t* /*tar*/, unsigned /*pos*/) -> int { return MTAR_ESUCCESS; };
  tar.close = [](mtar_t* /*tar*/) -> int { return MTAR_ESUCCESS; };

  valhalla::baldr::GraphReader reader(config.get_child("mjolnir"));
  mtar_header_t tar_header;
  int result = mtar_open(&tar, tar_path.c_str(), "r");
  if (result != MTAR_ESUCCESS) {
      throw std::runtime_error("Failed to open TAR file: " + tar_path);
  }
  std::string tile_name = tileIDtoPath(edge_id);
  result = mtar_find(&tar, tile_name.c_str(), &tar_header);
  if (result != MTAR_ESUCCESS) {
      mtar_close(&tar);
      throw std::runtime_error("Failed to find tile: " + tile_name + " in TAR archive");
  }
  /// rest of the code
}

std::string tileIDtoPath(const valhalla::baldr::GraphId& edge_id) {
    int tile_level = edge_id.level();
    int tile_id = edge_id.tileid();

    int part1 = (tile_id / 1000000) % 1000; // Extract the first 3 digits
    int part2 = (tile_id / 1000) % 1000; // Extract the middle 3 digits
    int part3 = tile_id % 1000; // Extract the last 3 digits

    std::ostringstream tile_path;
    tile_path << tile_level << "/"
              << std::setw(3) << std::setfill('0') << part1 << "/"
              << std::setw(3) << std::setfill('0') << part2 << "/"
              << std::setw(3) << std::setfill('0') << part3 << ".gph";

    return tile_path.str();
}
```
This code is derived from [vaalhalla_traffic_poc](https://github.com/alinmindroc/valhalla_traffic_poc/blob/main/valhalla_code_overwrites/src/mjolnir/valhalla_traffic_demo_utils.cc). The idea is to load the traffic tar to memory and access the tile within it using the provided `GraphId`.  

The issue is that the resulting value from `tileIDtoPath` is in the format of `7/004/194/303.gph` and as you may already know, there is no such thing as 7 level in the tile hierarchy.  I'm unsure why I'm getting 7 since I'm directly accessing `edge_id.level()` which means it might be the issue with the way I'm preparing data for `OfflineMatch` or it could mean I'm not parsing `GraphId` to the file name correctly in `tileIDtoPath`.  

Any help would be much appreciated!

**Comment by kevinkreiser** [comment]
id have to check to be 100% certain but i suspect that you are getting `kInvalidGraphId` which is set to a constant that gives a crazy graph id such as one on level 7. so the real question is where are you getting the graphid from in the first place? base on the above it looks like from map matching. it is true that map matching doesnt always give 100% success for the whole list of gps/location points. the match result is per gps/location edge info, those points which didnt find a match will have the invalid edge id that you are seeing. to make it really robust, i would first take a look at the matches that are failing and if you think it should/shouldnt fail. if the latter then you should play with the radius you use for map matching until the matching is working as good as you think it should be. once that is doen you'll still need to weed out parts of the match that didnt actually match (its going to happen some times with real data).
